README.md
# gulp-messenger
### gulp plugin for browser, command line notification and logging!

[![Build Status](https://travis-ci.org/mikeerickson/gulp-messenger.svg?branch=master)](https://travis-ci.org/mikeerickson/gulp-messenger) [![Dependency Status](https://gemnasium.com/mikeerickson/gulp-messenger.svg)](https://gemnasium.com/mikeerickson/gulp-messenger) [![npm](https://img.shields.io/npm/dm/localeval.svg)]() [![npm](https://img.shields.io/badge/mocha-passed-green.svg)]()

View on [npm](https://www.npmjs.com/package/gulp-messenger)


### Install with [npm](npmjs.org)

```sh
npm install gulp-messenger
```

### Usage

```js
var msg = require('gulp-messenger');

msg.init(); // initialize module, otherwise defaults will be used

// this example will enable file logging (see options below)
msg.init({logToFile: true});


// this will disable logging to file and interpolation variables will be font style as message (no bold)
var options = { logToFile: false, boldVariables: false };
msg.init(options);


// You can use titleCase or lower case for msg methods
msg.Info('-', 'Loading...', '*');
msg.log    ('style: <%= name %>', {name: 'info'});     // default color White
msg.Info   ('style: <%= name %>', {name: 'info'});     // default color Cyan
msg.Success('style: <%= name %>', {name: 'success'});  // default color Green
msg.Warning('style: <%= name %>', {name: 'warning'});  // default color Yellow
msg.Error  ('style: <%= name %>', {name: 'error'});    // default color Red
msg.Note   ('style: <%= name %>', {name: 'note'});     // default color Orange
msg.Time   ('style: <%= name %>', {name: 'time'});
msg.Debug  ('style: <%= name %>', {name: 'debug'});
msg.Table  (array|object)
```


### API
#### info/Info, success/Success, warning/Warning, error/Error, note/Note, time/Time, debug/Debug, line/Line

- lowercase methods are to be used in gulp pipeline or outside (same API can be used for both Node, Browser, and Gulp)
- Titlecase methods are to be used outside of gulp (ie Node or browser)

Default Options (supplied to `init` method)

```js
var defOptions = {
    logToFile:          false,
    logTimestampFormat: 'YYYY-MM-DD HH:mm:ss Z',
    logToConsole:       true,
    logPath:            'logs/',
    logFile:            'app.log',
    timestamp:          false,
    rotateLog:          false,
    boldVariables:      true,
    useDumpForObjects:  true
};

logToFile (default: false)
- will log the supplied message to `logger` instance and log to file

logToConsole (default: true)
- will log message to console

logTimestampFormat (default: YYYY-MM-DD HH:mm:ss Z)
- when logging to file, uses momentjs formats

logPath (default: 'logs/' at root level)
- desired path where log files will be stored

logFile (default: 'app.log')
- desired filename where log files will be stored

timestamp (default: false)
- Will include timestamp on consoled messages

rotateLog (default: false)
- will create a new log file each day

boldVariables (default: true)
- when perform interpolation, the actual variable will be bold

useDumpForObjects (default: true)
- when enabled, is passing a single object paramter, msg.dump() will be used instead of standard console output
- when disabled, native msg.xxx will be used (default colors, etc)


```

For example `info`
Use `msg.info` for each file into the stream

```js

msg.Info('This information message logged to console and optionally log file.  Configured by `options` parameter supplied to `msg.init` method');
msg.info('This information message logged to console and optionally log file.  Configured by `options` parameter supplied to `msg.init` method');

or -

// If you using in gulp stream, include the `flush` property as shown here
// All messenger routines (except dump) are supported in the gulp pipleline
gulp.src('src/**/*')
    .pipe(msg.flush.info('Piping Message'));  // 'Piping Message' for each file
```

Use `msg.flush.info` at the and of the stream

```js
gulp.src('src/**/*')
    .pipe(msg.flush.info('Process Completed Successfully')); //'Process Completed Successfully' at the and of stream
```

Use `msg.Info` from the outside of the stream

```js
msg.Info('Application Message'); //'Application Message' in node.js application
```

### msg.version|msg.Version

Returns current module version

### msg.Info([before,] message, [after,] [data])

Show message

#### Parameters

##### before
##### after
Type: `String`

Delimiter before/after the message. Each character is repeated 80 times

#### Usage

```js
msg.Info('--', 'Hello World', '*')
```

##### data
Type: `Object`

Data for message. Inherited values:

<pre>
env           - process.env
file          - vinyl file
file.relative - relative path (extra field)
file.basename - basename (extra field)
duration      - duration of streaming
totalDuration - duration from gulpfile start
</pre>


##### message
Type: `String`

Lodash compatible template (may also use Underscore over Lodash).

#### Usage

```js
msg.Info('Environment: <%= env.NODE_ENV %>. Name: <%= name %>', {name: 'codedungeon'})
//Environment: dev. name: codedungeon
```


### License

Copyright (c) 2015-2016 Mike Erickson
Released under the MIT license


### Credits

gulp-messenger written by Mike Erickson

E-Mail: [codedungeon@gmail.com](mailto:codedungeon@gmail.com)

Twitter: [@codedungeon](http://twitter.com/codedungeon)

Webiste: [codedungeon.org](http://codedungeon.org)

***


CHANGELOG.md
# gulp-messenger
### gulp plugin for browser, command line notification and logging!

[![Build Status](https://travis-ci.org/mikeerickson/gulp-messenger.svg?branch=master)](https://travis-ci.org/mikeerickson/gulp-messenger) [![Dependency Status](https://gemnasium.com/mikeerickson/gulp-messenger.svg)](https://gemnasium.com/mikeerickson/gulp-messenger) [![npm](https://img.shields.io/npm/dm/localeval.svg)]() [![npm](https://img.shields.io/badge/mocha-passed-green.svg)]()

## Changelog

- 0.28.0
  - Add check when create `logs/` directory to only create when `logToFile` is enabled
  
- 0.25.0
  - added .table method (output object in table)
  - uses `easyTable`
  
- 0.24.0
  - lots of little changes :-)

- 0.19.0
  - Added `version` method
  - Added `name` method

- 0.18.0
  - Updated to `reveal` and viewable on `npmaddict.com`

- 0.17.0
  - Add support for sprintf interpolation

- 0.14.0
  - Integrated chalkline source as lib instead of npm module
  - Added new option `chalklineBox` (default value `\u2584`)
    Example:
    `msg.init({chalklineBox: '\u2591'})`
  - Updated most of the dependencies (some need to stay at current version for compatability)
    - lodash-deep@1.6.0

- 0.13.0
    - Added `strict mode` check before enabling orange

- 0.12.0
    - Added `logTimestampFormat` option, accepts momentjs format
    - Default logTimestampFormat ( YYYY-MM-DD HH:mm:ss Z)

- 0.11.1
    - Fixed msg.note (ORANGE) color output to properly close color buffer when msg.note output displayed

- 0.11.9
    - Repurposed msg.note method to use default color of orange (color constant added in 0.10.0)
      msg.note('This will appear in orange');

- 0.10.0
    - Added new `orange` color which is default color for console.note routing
    - If you wish to use orange in any other output, the correct ASCI command to use

      var COLOR_ORANGE = "\033[38;5;214m";
      var COLOR_RESET  = "\033[m";
      msg.log(COLOR_ORANGE + 'Hello World!' + COLOR_RESET);

      * Note: The orange color is using 256 palette so it may not appear identical across !terminals
      * I have tested on OSX, Windows, CentOS 6/7 and Ubuntu 12/14

- 0.9.0
    - When executing `msg` command, supplying object as only parameter, it will use .dump command, displaying pretty output
      - If you dont want to use `dump` command for object output, disable `useDumpForObjects` options
    - Added mew `useDumpForObjects` option property to control default output option for simple objects
- 0.8.0
    - New command `msg.colors` which is an instance of Chalk
- 0.6.2
    - Added `dir` routine (calls `dump` internally)
    - Fixed issues related to passing object data in addition to messages (regression)
- 0.5.0
    - Lots of fixes internally before be pushed to public consumption
      - Added `chalk` instance ( msg.chalk... )
      - Added `chalkline` instance ( msg.chalkline... )
      - Added `dump` method ( msg.dump )
      - Updated `line` method

- 0.2.0
    - Fixed issues with properly display object literals
    - Added `version` method (returns current package version)

- 0.1.0
    - Added support for passing standard data as parameters 2..n
    - Added support for using `console.` instead of `msg` etc.
    - Enhanced package to support all `console` methods (eg console.log, console.error, etc)

- 0.0.43 Added new `header` entry point
    - Added msg.header method
      Displays message default text underline (doesn't have its own color)

- 0.0.42 Added new `log` entry point
    - Added msg.log method

- 0.0.41 Bug Fixes
    - Fixed issue with line printing feature (was not actually use line)
    - Refreshed all plugin dependencies


- 0.0.40 Regression Fixes
    - Fixed regression issue introduced in 0.0.30 causing logs with ojbect as message parameter

- 0.0.35 Bug Fixes

- 0.0.34
    - Internal (development level) additions
      * Todo and Test tasks support option of opening report file when completed

- 0.0.33
    - Workflow Refinements
      * gulpfile, tasks, etc

- 0.0.32
    - exposed msg.line method
      msg.line({character, numChars});
       * character (default =)
       * numChars  (default 80)

- 0.0.31
    - Fixed results when using msg.Time/time.  It was not echoing current time as it should

- 0.0.30 Logging To File
    - Fixed issue when logging to `logger` file instance

- 0.0.28 Bug Fixes
    - Fixed issue with `timestamp` option not properlly including timestamp

- 0.0.27 Bug Fix
    - Fixed a regression issue when using messenger commands in gulp pipeline

- 0.0.24 Small Fixes
    - fixed issue when supplied message is an 'object' causing an error when parsing tokens

- 0.0.22 Test Refactoring
    - Refactored tests

- 0.0.21 Added options
    - Added option `boldVariables` to set display of merged variables
    - Added option `logToConsole` to control if messages are sent to console
    - Added option `logToFile` to control is messages are sent to log
    - Added option `rotateLog` to control log files to have current date appended to filename

- 0.0.20 Added .debug command
    - Added msg.debug (msg.Debug) and supported text color


index.js
/*global require */

//  GULP MESSENGER
// =============================================================================

// this is required to support .orange color
// eventually want to restore "use strict" waiting on chalk

var VERSION      = require('./package.json').version;

var chalk        = require('chalk');
var is           = require('is_js');
var mkdirp       = require('mkdirp');
var moment       = require('moment');
var path         = require("path");
var prettyHrtime = require('pretty-hrtime');
var through      = require('through2');
var Table        = require('easy-table');
var Purdy        = require('purdy');
var bowser       = require('bowser');
var sprintf      = require("sprintf-js").sprintf;
var _            = require('lodash');
_.mixin(require('lodash-deep'));

var winston      = null;

if(bowser.name.length === 0) {
  var winston      = require('winston');
}

// MODULE CONSTANTS
// =============================================================================

var VALUE_REGEXP        = /<%=\s*([^\s]+)\s*%>/g;
var COLOR_ORANGE        = chalk.red();
var COLOR_RESET         = chalk.white();
var COLOR_CODES_REGEXP  = /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g;




/*eslint-disable */
if(! isStrictMode() ) {
  COLOR_ORANGE          = '\033[38;5;214m';
  COLOR_RESET           = '\033[m';
}
/*eslint-enable */

function isStrictMode() {
  return (typeof this === 'undefined');
}

// SETUP DEFAULT OPTIONS
// =============================================================================

var defOptions = {
  logToFile:          false,
  logToConsole:       true,
  logTimestampFormat: 'YYYY-MM-DD HH:mm:ss Z',
  logPath:            'logs/',
  logFile:            'app.log',
  timestamp:          false,
  rotateLog:          false,
  boldVariables:      true,
  showPipeFile:       true,
  useDumpForObjects:  true,
  lineLength:         80,
  chalklineBox:       '\u2584' // '\u2588'
};

// unicode blocks -- http://graphemica.com/blocks/block-elements
// SETUP CHALKLINE
// chalkline.options({block: defOptions.chalklineBox});


// SETUP WINSTON
// =============================================================================
// Initialize logger, additional settings will be created in `init` method

var logger = null;
if ( bowser.name.length === 0) {
  var logger = new (winston.Logger)({level: 'debug'});
}

// MODULE DEFINITION (used by module.exports below)
// =============================================================================

var messenger = {
  orange: function(msg) {
    if( !is.undefined(msg) ) {
      return COLOR_ORANGE + msg + COLOR_RESET;
    }
    return COLOR_ORANGE;
  },
  init:       init(),
  setOptions: setOptions(),
  name:       'gulp-messenger',
  info:       new Message('info'),
  log:        new Message('info'),
  success:    new Message('success'),
  warning:    new Message('warning'),
  warn:       new Message('warning'),
  error:      new Message('error'),
  note:       new Message('note'),
  table:      new Message('table'),
  time:       new Message('time'),
  debug:      new Message('debug'),
  line:       new Message('line'),
  header:     new Message('header'),
  dir:        function() {
    Purdy.apply(Purdy, arguments);
  },
  purdy:      function() {
    Purdy.apply(Purdy, arguments);
  },
  dump:       function() {
    Purdy.apply(Purdy, arguments);
  },
  version:    function() {
    return VERSION;
  },
  chalkline:  false,
  chalk:      chalk,
  colors:     chalk,
  flush: {
    info:    message('info', true),
    log:     message('info', true),
    success: message('success', true),
    warning: message('warning', true),
    warn:    message('warning', true),
    error:   message('error', true),
    note:    message('note', true),
    table:   message('table', true),
    time:    message('time', true),
    debug:   message('debug', true),
    line:    message('line', true),
    header:  message('header',true),
    chalkline: false,
    dir: function() {
      Purdy.apply(Purdy, arguments);
    },
    purdy: function() {
      Purdy.apply(Purdy, arguments);
    },
    dump: function(){
      Purdy.apply(Purdy, arguments);
    }
  },
  Info:       new Message('info'),
  Log:        new Message('info'),
  Success:    new Message('success'),
  Warning:    new Message('warning'),
  Warn:       new Message('warning'),
  Error:      new Message('error'),
  Note:       new Message('note'),
  Time:       new Message('time'),
  Debug:      new Message('debug'),
  Line:       new Message('line'),
  Header:     new Message('header'),
  Table:      new Message('table'),
  Purdy:     function() {
    Purdy.apply(Purdy, arguments);
  },
  Dump:      function() {
    Purdy.apply(Purdy, arguments);
  },
  Version:   function() {
    return VERSION;
  }
};

// MODULE ENTRY POINTS
// =============================================================================

function message(style, useFlush) {
  var totalStart    = process.hrtime();

  return function() {
    var args     = getArgs(arguments);
    var lastFile = {};
    var start    = process.hrtime();

    function transform(file, enc, callback) {
      lastFile = file;

      if (!useFlush) {
        args.data.file          = _.clone(file);
        args.data.file.relative = path.relative(file.base, file.path);
        args.data.file.basename = path.basename(file.path);
        args.data.duration      = prettyHrtime(process.hrtime(start));
        args.data.totalDuration = prettyHrtime(process.hrtime(totalStart));
        notify(style, args.before, args.message, args.after, args.data);
      }
      callback(null, file);
    }

    function flush(callback) {
      if (useFlush) {

        // not sure how this is going to be used as we are clearing .data below
        args.data.file          = _.clone(lastFile);
        args.data.duration      = prettyHrtime(process.hrtime(start));
        args.data.totalDuration = prettyHrtime(process.hrtime(totalStart));

        // gulp pipeline specific
        defOptions.timestamp = true; // timestamp always true when in pipeline
        args.data = {};              // no data available in gulp pipeline

        // this is only appropriate adjustment to message (to include the filename)
        var msg = args.message;
        if (defOptions.showPipeFile) {
          msg += ' [' + lastFile.relative + '] ';
        }

        notify(style, args.before, msg, args.after, args.data);
      }
      callback();
    }

    return through.obj(transform, flush);
  };
}

function Message(style) {
  return function() {
    var args = getArgs(arguments);
    if((is.object(args.message)) && (arguments.length === 1) && (defOptions.useDumpForObjects)) {
      /* jshint -W064 */
      if(style === 'table') {
        console.log(Table.print(args.message));
      }
      else {
        Purdy(args.message);
      }
    } else {
      notify(style, args.before, args.message, args.after, args.data);
    }
  };

}


// MODULE PRIVATE METHOD
// =============================================================================

function init(options) {

  var added = false;

  return function(options) {

    var isBrowser = bowser.name.length > 0;

    if(is.not.undefined(options)) {
      defOptions = _.defaults(options, defOptions);
    }

    if(defOptions.logPath[defOptions.logPath.length] !== '/') {
      defOptions.logPath += '/';
    }

    if(! isBrowser) {
      // create log path if it doesn't already exist, only create if we have enabled fileLogging
      if (defOptions.logToFile)
        mkdirp(defOptions.logPath);

      defOptions.logFilename = defOptions.logPath + defOptions.logFile;
      if ( defOptions.rotateLog ) {
        logger.add(winston.transports.DailyRotateFile,{
          filename: defOptions.logFilename,
          timestamp: function() {
            return moment().format(defOptions.logTimestampFormat);
          }
        });
      } else {
        if( ! added ) {
          added = true;
          logger.add(winston.transports.File,{
            filename: defOptions.logFilename,
            timestamp: function() {
              return moment().format(defOptions.logTimestampFormat);
            }
          });
        }
      }
    } else {
      console.warn('Logging Disabled On Browser');
    }

    // only do this once regardless so outside browser check
    defOptions.logInitialized = true;

    // update chalkine in the event supplied block or length
    // chalkline.options({block: defOptions.chalklineBox});

  };
}

function notify(style, before, message, after, data) {

debugger;

  var text, variable;
  var result = '';
  var tokens;

  // 2015.05.28 - added bounds check, exposed when adding *.line() routine
  if( is.undefined(message) ) { message = ''; }

  debugger;

  if ( (is.not.object(message)) && (is.not.number(message)) ) {
    tokens = message.split(VALUE_REGEXP);
  } else {
    tokens = message;
  }

  switch (style) {
    case "info":
    case 'log':
      text     = chalk.cyan;
      variable = chalk.cyan.bold;
      break;
    case "success":
      text     = chalk.green;
      variable = chalk.green.bold;
      break;
    case "warning":
      text     = chalk.yellow;
      variable = chalk.yellow.bold;
      break;
    case "error":
      text     = chalk.red;
      variable = chalk.red.bold;
      break;
    case "note":
      text     = COLOR_ORANGE;
      variable = chalk.white;
      break;
    case "table":
      text     = chalk.white;
      variable = chalk.white;
    case "time":
      text     = chalk.magenta;
      variable = chalk.magenta.bold;
      break;
    case "debug":
      text     = chalk.grey.dim;
      variable = chalk.grey.dim.bold;
      break;
    case "line":
      text     = chalk.green;
      variable = chalk.green.bold;
      break;
    case "header":
      text     = chalk.white.underline;
      variable = chalk.white;
      break;
    default:
      text     = chalk.gray;
      variable = chalk.white;
      break;
  }

// if we don't have bold variables (for merging), set variable to text color
  if( ! defOptions.boldVariables ) {
    variable = text;
    if ( text === chalk.gray ) {
      variable = chalk.white;
    }
  }

  for (var i = 0; i < tokens.length; i++) {
    if (i%2) {
      result += variable(_.deepGet(data, tokens[i]) || '');
    } else {
      if ( text === COLOR_ORANGE) {
        result = text + tokens[i] + COLOR_RESET;
      } else {
        result += text(tokens[i] || '');
      }
    }
  }

// if the supplied message is an object, return object (string) as result
  if(! result.length ) {
    result = tokens;
  }

  function setConsole(data) {

    var callData = {};
    if ( is.not.undefined(data) ) {
      callData = data;
    }
    var hCurrentTime = moment().format('HH:mm:ss');

    if (( defOptions.logToConsole ) && ( style !== 'line')){
      if ( defOptions.timestamp || (style === 'time')) {
        if ( style === 'time') {
          console.log('[' + chalk.grey(hCurrentTime) + '] ' + hCurrentTime);
        } else {
          if ( result ) {
            if( Object.keys(arguments[0]).length === 2 ) {
              console.log('[' + chalk.grey(hCurrentTime) + '] ' + result, arguments[0]);
            } else {
              console.log('[' + chalk.grey(hCurrentTime) + '] ' + result);
            }
          }
        }
      } else {

        if ( is.object(callData)) {
          if ( Object.keys(callData).length > 0 ) {
            if(is.array(callData)) {
              callData.splice(0,0,result);
              console.log.apply(console, callData);
            }
          } else {
            console.log(result);
          }
        } else {
          console.log(result, callData);
        }
      }
    }

    // if we are outputting a line, just spit out what we got using `lineLength`
    if ( style === 'line') {
      setLine(result);
    }

  }

  function setLine(line) {

    if (!line) { return; }

    var result = '';
    for (var i = 0; i < defOptions.lineLength; i++) {
      result += line;
    }
    if ( defOptions.logToConsole ) {
      // we have to handle orange separately
      if(typeof(text) === 'function') {
        console.log(text(result));
      } else {
        console.log(COLOR_ORANGE + result + COLOR_RESET);
      }
    }
  }

  function logToFile(style, result) {

    // make sure we received text, otherwise bail
    if ( ! is.string(result) ) { return; }

    // don't bother logging if we have no message
    if ( result.length === 0  ) { return; }

    // strip out all the color codes, etc. from message before logging
    var msg = result.replace(COLOR_CODES_REGEXP,'');

    // secondary check to make sure we have logging enabled
    if (defOptions.logToFile) {
      switch (style) {
        case 'error':
          logger.error(msg); break;
        case 'warning':
        case 'warn':
          logger.warn(msg); break;
        case 'success':
          logger.info(msg); break;
        case 'debug':
          logger.log('debug', msg); break;
        case 'note':
        case 'info':
        case 'log':
        case 'default':
          logger.info(msg); break;
        case 'table':
          return false;
      }
    }

  }

  setLine(before);
  setConsole(data);
  setLine(after);

  // don't bother logging if it is disabled, save some processor fumes
  (defOptions.logToFile) ? logToFile(style, result) : '';

}

function getArgs(args) {

  var msg = '';
  if(args.length > 0 ) {
    var msg = args[0];
    if(is.string(msg)) {
      if ( msg.indexOf('%s') > 0 ) {
        // we are processing via sprintf
        var params = Array.prototype.slice.call(args);
        params.shift();
        return {
          before: '',
          message: msg,
          after: '',
          data: params
        }
      }
    }
  }
  var result = {
    before:  args[0],
    message: args[1],
    after:   args[2] || null,
    data:    args[3] || {}
  };

  if(args.length === 1) {
    result.before = '';
    result.message = args[0];
    result.after = '';
  }

  if(args.length === 2) {
    result.before = args[0];
    result.message = args[1];
    result.after = '';
  }

  if( is.not.undefined(args[1]) ) {
    if(is.not.string(args[1])) {
      result.before  = null;
      result.message = args[0];
      result.after   = null;
      result.data    = args[1];
      // result.data    = args;
    } else if (is.not.string(args[2])) {
      result.before  = args[0];
      result.message = args[1];
      result.after   = null;
      result.data    = args[2];
    }
  }
  if( is.undefined(result.data) ) {
    result.data = {};
    result.data.file = '';
  }

  //result.data = _.merge({env: process.env}, result.data);
  return result;
}

function setOptions(options) {

  var added = true;

  return function(options) {

    if(is.not.undefined(options)) {
      defOptions = _.defaults(options, defOptions);
    }

    if(defOptions.logPath[defOptions.logPath.length] !== '/') {
      defOptions.logPath += '/';
    }

    // create log path if it doesn't already exist
    mkdirp(defOptions.logPath);

    defOptions.logFilename = defOptions.logPath + defOptions.logFile;
    if ( defOptions.rotateLog ) {
      logger.add(winston.transports.DailyRotateFile,{filename: defOptions.logFilename});
    } else {
      if( !added ) {
        added = true;
        logger.add(winston.transports.File,{filename: defOptions.logFilename});
      }
    }
    defOptions.logInitialized = true;
  };
}

// MODULE EXPORT
// =============================================================================

module.exports = messenger;
